kmir/k-src/mir-rvalue.md

```diff
@@ -0,0 +1,371 @@
+      orBool OP_NAME ==String "Offset"
+  rule isBinOp(_)       => false [owise]
+
+  syntax CheckedBinaryOpName ::= "CheckedAdd"
+                               | "CheckedSub"
+                               | "CheckedMul"
+                               | "CheckedShl"
+                               | "CheckedShr"
+
+endmodule
+```
+
+Evaluation of rvalues
+---------------------
+
+```k
+module MIR-RVALUE
+  imports MIR-RVALUE-SYNTAX
+  imports MIR-BUILTINS-SYNTAX
+  imports MIR-TYPES
+  imports MIR-CONFIGURATION
+```
+
+Evaluate a syntactic `RValue` into a semantics `RValueResult`. Inspired by [eval_rvalue_into_place](https://github.com/rust-lang/rust/blob/bd43458d4c2a01af55f7032f7c47d7c8fecfe560/compiler/rustc_const_eval/src/interpret/step.rs#L148).
+
+```k
+  syntax InterpResult ::= evalRValue(FunctionLikeKey, RValue) [function]
+  //--------------------------------------------------------------------
+  rule evalRValue(FN_KEY, VALUE:Operand)   => evalOperand(FN_KEY, VALUE)
+  rule evalRValue(FN_KEY, UN_OP:UnaryOp)   => evalUnaryOp(FN_KEY, UN_OP)
+  rule evalRValue(FN_KEY, BIN_OP:BinaryOp) => evalBinaryOp(FN_KEY, BIN_OP)
+  rule evalRValue(_FN_KEY, RVALUE)         => Unsupported(RVALUE) [owise]
+```
+
+### `Operand` evaluation
+
+```k
+  syntax MIRValue ::= evalOperand(FunctionLikeKey, Operand) [function]
+  //------------------------------------------------------------------
+  rule evalOperand(_, const VALUE:ConstantValue)     => evalConstantValue(VALUE)
+  rule evalOperand(FN_KEY, LOCAL:Local)                   => evalLocal(FN_KEY, LOCAL)
+  rule evalOperand(FN_KEY, move LOCAL:Local)              => evalLocal(FN_KEY, LOCAL)
+```
+
+### `UnaryOp` evaluation
+
+```k
+  syntax MIRValue ::= evalUnaryOp(FunctionLikeKey, UnaryOp) [function]
+  //------------------------------------------------------------------
+  rule evalUnaryOp(FN_KEY, NAME:IdentifierToken (X:Operand)) =>
+       evalUnaryOpImpl(FN_KEY, String2UnaryOpName(IdentifierToken2String(NAME)), X)
+
+  syntax MIRValue ::= evalUnaryOpImpl(FunctionLikeKey, UnaryOpName, Operand) [function]
+  //-----------------------------------------------------------------------------------
+  rule evalUnaryOpImpl(FN_KEY, Not, X)    => notBool {evalOperand(FN_KEY, X)}:>Bool
+  rule evalUnaryOpImpl(FN_KEY, Neg, X)    => 0 -Int {evalOperand(FN_KEY, X)}:>Int
+```
+
+### `BinaryOp` evaluation
+
+```k
+  syntax MIRValue ::= evalBinaryOp(FunctionLikeKey, BinaryOp) [function]
+  //--------------------------------------------------------------------
+  rule evalBinaryOp(FN_KEY, NAME:IdentifierToken (X:Operand, Y:Operand)) =>
+       evalBinaryOpImpl(FN_KEY, String2BinaryOpName(IdentifierToken2String(NAME)), X, Y)
+
+  syntax MIRValue ::= evalBinaryOpImpl(FunctionLikeKey, BinaryOpName, Operand, Operand) [function]
+  //-----------------------------------------------------------------------
+  rule evalBinaryOpImpl(FN_KEY, Add, X, Y)    => {evalOperand(FN_KEY, X)}:>Int +Int {evalOperand(FN_KEY, Y)}:>Int
```

Comment:
No, the results of `evalOperand` are not castable to `Int`  and the execution of `evalBinaryOpImpl` will get stuck if it encounters arguments of the invalid type.

However, it is highly unlikely that such invalidly-typed arguments will ever be supplied, since we assume that the Mir is produces by `rustc` which does typecheking prior to Mir emission. Considering this, I made the call to go for the simplest possible semantics to build a prototype quickly.